server/application/application.go

```diff
@@ -1892,6 +1893,51 @@ func (s *Server) Rollback(ctx context.Context, rollbackReq *application.Applicat
 	return a, nil
 }
 
+func (s *Server) ListLinks(ctx context.Context, req *application.ListLinksRequest) (*application.LinksResponse, error) {
+	appName := req.GetName()
+	appNs := s.appNamespaceOrDefault(req.GetNamespace())
+	a, err := s.appclientset.ArgoprojV1alpha1().Applications(appNs).Get(ctx, appName, metav1.GetOptions{})
```

Comment:
Should there be an RBAC check to make sure the user has access to this application?

If so, the frontend should send the project name as well so that we can perform the RBAC check before getting the application.

Reply:
I think those checks are poorly designed. They rely on fetching `a` (the application spec) before performing the RBAC check, which partially defeats the point of doing the RBAC check at all (an unauthorized user can spam the k8s API).

The better way to do the check is:

1) require the `appName`, `appNamespace`, and `project` parameters
2) do an initial RBAC check using those parameters - if the check fails, return a generic "unauthorized" error
3) actually get the Application manifest
4) check that the actual project name from the application manifest matches the project name from the request parameters - if they're not equal, return the same generic "unauthorized" error as from step 2

It's important that both error messages are precisely the same, because otherwise a malicious user could infer the existence of an app outside their access boundaries by spamming the API server with different parameters.

Reply:
Yep, there's a GitHub issue to track fixing the other services.

I actually don't think there's an Application endpoint that does it correctly at this point. 😬

Reply:
@crenshaw-dev have updated the PR with the rbac check, ptal

Reply:
lgtm! I really like that the request is fully parameterized. No pre-RBAC requests.  :-)

Reply:
I understand specifying the project makes it security-friendly but API became inconsistent. We have 5~6 more APIs with exact same issue (possible app enumeration vulnerability). 

I propose to remove project field and address this security concern in a separate PR. I'm pretty sure adding one more API that potentially allow to enumerate apps is not making situation any worse